kie-server-parent/kie-server-tests/kie-server-integ-tests-optaplanner/src/test/java/org/kie/server/integrationtests/optaplanner/OptaplannerIntegrationTest.java

```diff
@@ -0,0 +1,326 @@
+        instance.setSolverConfigFile( SOLVER_1_CONFIG );
+        assertSuccess( client.createContainer( CONTAINER_1_ID, new KieContainerResource( CONTAINER_1_ID, kjar1 ) ) );
+
+        assertSuccess( solverClient.createSolver( CONTAINER_1_ID, SOLVER_1_ID, instance ) );
+
+        // the following status starts the solver
+        instance.setStatus( SolverInstance.SolverStatus.SOLVING );
+        instance.setPlanningProblem( loadPlanningProblem( 10, 100 ) );
+        ServiceResponse<SolverInstance> response = solverClient.updateSolverState( CONTAINER_1_ID, SOLVER_1_ID, instance );
+        assertSuccess( response );
+        assertEquals( SolverInstance.SolverStatus.SOLVING, response.getResult().getStatus() );
+
+        ServiceResponse<Solution> solutionResponse = solverClient.getSolverBestSolution(CONTAINER_1_ID, SOLVER_1_ID);
+        assertSuccess( solutionResponse );
+
+        // TODO evaluate best solution
+
+        assertSuccess( solverClient.disposeSolver( CONTAINER_1_ID, SOLVER_1_ID ) );
+    }
+
+    @Test
+    public void testGetBestSolutionNotExistingSolver() throws Exception {
+        assertSuccess( client.createContainer( CONTAINER_1_ID, new KieContainerResource( CONTAINER_1_ID, kjar1 ) ) );
+
+        try {
+            solverClient.getSolverBestSolution(CONTAINER_1_ID, SOLVER_1_ID);
+        } catch (KieServicesException e) {
+            assertResultContainsStringRegex(e.getMessage(), ".*Solver.*not found in container.*");
+        }
+    }
+
+    @Test
+    public void testTerminateEarlyNotExistingSolver() throws Exception {
+        assertSuccess( client.createContainer( CONTAINER_1_ID, new KieContainerResource( CONTAINER_1_ID, kjar1 ) ) );
+
+        SolverInstance instance = new SolverInstance();
+        instance.setSolverConfigFile( SOLVER_1_CONFIG );
+        instance.setStatus( SolverInstance.SolverStatus.NOT_SOLVING );
+        instance.setPlanningProblem( loadPlanningProblem( 5, 10 ) );
+        try {
+            solverClient.updateSolverState( CONTAINER_1_ID, SOLVER_1_ID, instance );
+        } catch (KieServicesException e) {
+            assertResultContainsStringRegex(e.getMessage(), ".*Solver.*not found in container.*");
+        }
+    }
+
+    @Test
+    public void testTerminateEarlyStoppedSolver() throws Exception {
+        SolverInstance instance = new SolverInstance();
+        instance.setSolverConfigFile( SOLVER_1_CONFIG );
+        assertSuccess( client.createContainer( CONTAINER_1_ID, new KieContainerResource( CONTAINER_1_ID, kjar1 ) ) );
+        assertSuccess( solverClient.createSolver( CONTAINER_1_ID, SOLVER_1_ID, instance ) );
+
+        instance.setStatus( SolverInstance.SolverStatus.NOT_SOLVING );
+        instance.setPlanningProblem( loadPlanningProblem( 50, 300 ) );
+        assertSuccess( solverClient.updateSolverState( CONTAINER_1_ID, SOLVER_1_ID, instance ) );
+    }
+
+    @Test @Ignore("Ignoring for now. Geoffrey will review why this is failing.")
```

Comment:
I think this is failing because solver is started asynchronously:
Method updateSolverState in SolverServiceBase starts solver in new thread and then update state of SolverInstanceContext in updateSolverStatus. There is condition solver.isSolving(), but result of that condition depends on whether thread with solver already started or not.